### ISSUE TITLE
2.106: test suite is failing when `IO-Compress` is build against zlib-ng

### ISSUE THREAD

**Issue by kloczek** [pmqs/IO-Compress#46]
Looks like there are few issues in test suite
```consoole
+ cd IO-Compress-2.106
+ /usr/bin/make -O -j48 V=1 VERBOSE=1 test COMPRESS_ZLIB_RUN_ALL=1
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/000prereq.t ................... ok
t/001bzip2.t .................... ok
t/001zlib-generic-deflate.t ..... ok
t/001zlib-generic-gzip.t ........ ok
t/001zlib-generic-rawdeflate.t .. ok
t/001zlib-generic-zip.t ......... ok
t/002any-deflate.t .............. ok
t/002any-gzip.t ................. ok
t/002any-rawdeflate.t ........... ok
t/002any-transparent.t .......... ok
t/002any-zip.t .................. ok
t/004gziphdr.t .................. ok

#     Failed test (t/005defhdr.t at line 184)
#          got: '7'
#     expected: '5'

#     Failed test (t/005defhdr.t at line 187)
#          got: '1'
#     expected: '9'

#     Failed test (t/005defhdr.t at line 184)
#          got: '7'
#     expected: '5'

#     Failed test (t/005defhdr.t at line 187)
#          got: '1'
#     expected: '9'
# Looks like you failed 4 tests of 596.
t/005defhdr.t ...................
Dubious, test returned 4 (wstat 1024, 0x400)
Failed 4/596 subtests
t/006zip.t ...................... ok
t/010examples-bzip2.t ........... ok
t/010examples-zlib.t ............ ok
t/011-streamzip.t ............... ok
t/01misc.t ...................... ok
t/020isize.t .................... ok
t/050interop-gzip.t ............. ok
t/100generic-bzip2.t ............ ok
t/100generic-deflate.t .......... ok
t/100generic-gzip.t ............. ok
t/100generic-rawdeflate.t ....... ok
t/100generic-zip.t .............. ok
t/101truncate-bzip2.t ........... ok
t/101truncate-deflate.t ......... ok
t/101truncate-gzip.t ............ ok
# Looks like you planned 626 tests but only ran 616.
t/101truncate-rawdeflate.t ......
Dubious, test returned 10 (wstat 2560, 0xa00)
Failed 10/626 subtests
t/101truncate-zip.t ............. ok
t/102tied-bzip2.t ............... ok
t/102tied-deflate.t ............. ok
t/102tied-gzip.t ................ ok
t/102tied-rawdeflate.t .......... ok
t/102tied-zip.t ................. ok
t/103newtied-bzip2.t ............ ok
t/103newtied-deflate.t .......... ok
t/103newtied-gzip.t ............. ok
t/103newtied-rawdeflate.t ....... ok
t/103newtied-zip.t .............. ok
t/104destroy-bzip2.t ............ ok
t/104destroy-deflate.t .......... ok
t/104destroy-gzip.t ............. ok
t/104destroy-rawdeflate.t ....... ok
t/104destroy-zip.t .............. ok
t/105oneshot-bzip2.t ............ ok
t/105oneshot-deflate.t .......... ok
t/105oneshot-gzip-only.t ........ ok
t/105oneshot-gzip.t ............. ok
t/105oneshot-rawdeflate.t ....... ok
t/105oneshot-zip-bzip2-only.t ... ok
t/105oneshot-zip-only.t ......... ok
t/105oneshot-zip-store-only.t ... ok
t/105oneshot-zip.t .............. ok
t/106prime-bzip2.t .............. ok
t/106prime-deflate.t ............ ok
t/106prime-gzip.t ............... ok
t/106prime-rawdeflate.t ......... ok
t/106prime-zip.t ................ ok
t/107multi-bzip2.t .............. ok
t/107multi-deflate.t ............ ok
t/107multi-gzip.t ............... ok
t/107multi-rawdeflate.t ......... ok
t/107multi-zip-only.t ........... ok
t/107multi-zip.t ................ ok
t/108anyunc-bzip2.t ............. ok
t/108anyunc-deflate.t ........... ok
t/108anyunc-gzip.t .............. ok
t/108anyunc-rawdeflate.t ........ ok
t/108anyunc-transparent.t ....... ok
t/108anyunc-zip.t ............... ok
t/109merge-deflate.t ............ ok
t/109merge-gzip.t ............... ok
t/109merge-rawdeflate.t ......... ok
t/109merge-zip.t ................ skipped: not implemented yet
t/110encode-bzip2.t ............. ok
t/110encode-deflate.t ........... ok
t/110encode-gzip.t .............. ok
t/110encode-rawdeflate.t ........ ok
t/110encode-zip.t ............... ok
t/111const-deflate.t ............ ok
t/112utf8-zip.t ................. ok
t/113issues.t ................... ok
t/999meta-json.t ................ skipped: Test::CPAN::Meta::JSON required for testing META.json
t/999meta-yml.t ................. skipped: Test::CPAN::Meta required for testing META.yml
t/999pod.t ...................... ok

#     Failed test (t/cz-01version.t at line 34)
#          got: '1.2.11'
#     expected: '1.2.11.zlib-ng'
#
# The version of zlib.h does not match the version of libz
#
# You have zlib.h version 1.2.11
#        and libz   version 1.2.11.zlib-ng
#
# You probably have two versions of zlib installed on your system.
# Try removing the one you don't want to use and rebuild.
# Looks like you failed 1 test of 3.
t/cz-01version.t ................
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/3 subtests

#     Failed test (t/cz-03zlib-v1.t at line 55)
#          got: '1.2.11.zlib-ng'
#     expected: '1.2.11'

#     Failed test (t/cz-03zlib-v1.t at line 712)
#          got: 1
#     expected: -3
# Looks like you failed 2 tests of 472.
t/cz-03zlib-v1.t ................
Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/472 subtests
t/cz-05examples.t ............... ok

#     Failed test (t/cz-06gzsetp.t at line 48)
#          got: '1.2.11.zlib-ng'
#     expected: '1.2.11'
# Looks like you failed 1 test of 52.
t/cz-06gzsetp.t .................
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/52 subtests

#     Failed test (t/cz-08encoding.t at line 43)
#          got: '1.2.11.zlib-ng'
#     expected: '1.2.11'
# Looks like you failed 1 test of 30.
t/cz-08encoding.t ...............
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/30 subtests

#     Failed test (t/cz-14gzopen.t at line 34)
#          got: '1.2.11.zlib-ng'
#     expected: '1.2.11'
# Looks like you failed 1 test of 265.
t/cz-14gzopen.t .................
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/265 subtests
t/globmapper.t .................. ok

Test Summary Report
-------------------
t/005defhdr.t                 (Wstat: 1024 Tests: 596 Failed: 4)
  Failed tests:  75, 78, 375, 378
  Non-zero exit status: 4
t/101truncate-rawdeflate.t    (Wstat: 2560 Tests: 616 Failed: 0)
  Non-zero exit status: 10
  Parse errors: Bad plan.  You planned 626 tests but ran 616.
t/cz-01version.t              (Wstat: 256 Tests: 3 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
t/cz-03zlib-v1.t              (Wstat: 512 Tests: 472 Failed: 2)
  Failed tests:  4, 196
  Non-zero exit status: 2
t/cz-06gzsetp.t               (Wstat: 256 Tests: 52 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
t/cz-08encoding.t             (Wstat: 256 Tests: 30 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
t/cz-14gzopen.t               (Wstat: 256 Tests: 265 Failed: 1)
  Failed test:  3
  Non-zero exit status: 1
Files=93, Tests=69154, 992 wallclock secs ( 5.82 usr  0.47 sys + 968.18 cusr  2.37 csys = 976.84 CPU)
Result: FAIL
Failed 7/93 test programs. 10/69154 subtests failed.
make: *** [Makefile:851: test_dynamic] Error 255
```

**Comment by pmqs** [comment]
Thanks @kloczek - just to be clear -- are you are using zlib-ng from https://github.com/zlib-ng/zlib-ng ?

**Comment by kloczek** [comment]
Yep 😄

**Comment by pmqs** [comment]
Hey @kloczek 

your test output shows that `IO::Compress` and `Compress::Raw::Zlib` (which is ultimately what interfaces to the `zlib` library) were built with a standard copy of `zlib` version `1.2.11`.

You have then either replaced the system copy of `libz` with a `libz-ng` version or used a linker option, like `LD_PRELOAD`, to force the use of the `ng` version. 

Is that the case?

It means that `IO::Compress` and `Compress::Raw::Zlib`  were compiled using a standard `zlib` copy of  the `zlib.h` header file, but at runtime you are using a different version of the library, namely the one from `zlib-ng`.  There is no guarantee that the `zlib` interface contained in `zlib.h` , which gets used at compile time, matches the `zlib-ng` library used at runtime.

I don't support building with `zlib-ng` at the moment, but it shouldn't be that difficult to get it working. You are the first person to raise it. I've created an enhancement ticket  to add support for `zlib-ng` here -- https://github.com/pmqs/Compress-Raw-Zlib/issues/9.

**Comment by pmqs** [comment]
Just found the related issue https://github.com/zlib-ng/zlib-ng/issues/1234

**Comment by pmqs** [comment]
Fixed in 2.201
